src/machine/machine_k210.go

```diff
@@ -0,0 +1,642 @@
+	errI2CTxAbort               = errors.New("I2C transmition has been aborted.")
+)
+
+func (p Pin) setFPIOAIOPull(pull fpioaPullMode) {
+	switch pull {
+	case fpioaPullNone:
+		kendryte.FPIOA.IO[uint8(p)].ClearBits(kendryte.FPIOA_IO_PU & kendryte.FPIOA_IO_PD)
+	case fpioaPullUp:
+		kendryte.FPIOA.IO[uint8(p)].SetBits(kendryte.FPIOA_IO_PU)
+		kendryte.FPIOA.IO[uint8(p)].ClearBits(kendryte.FPIOA_IO_PD)
+	case fpioaPullDown:
+		kendryte.FPIOA.IO[uint8(p)].ClearBits(kendryte.FPIOA_IO_PU)
+		kendryte.FPIOA.IO[uint8(p)].SetBits(kendryte.FPIOA_IO_PD)
+	}
+}
+
+// SetFPIOAFunction is used to configure the pin for one of the FPIOA functions.
+// Each pin on the Kendryte K210 can be configured with any of the available FPIOA functions.
+func (p Pin) SetFPIOAFunction(f FPIOAFunction) {
+	kendryte.FPIOA.IO[uint8(p)].Set(fpioaFuncDefaults[uint8(f)])
+}
+
+// FPIOAFunction returns the current FPIOA function of the pin.
+func (p Pin) FPIOAFunction() FPIOAFunction {
+	return FPIOAFunction((kendryte.FPIOA.IO[uint8(p)].Get() & kendryte.FPIOA_IO_CH_SEL_Msk))
+}
+
+// Configure this pin with the given configuration.
+// The pin must already be set as GPIO or GPIOHS pin.
+func (p Pin) Configure(config PinConfig) {
+	var input bool
+
+	// Check if the current pin's FPIOA function is either GPIO or GPIOHS.
+	f := p.FPIOAFunction()
+	if f < FUNC_GPIOHS0 || f > FUNC_GPIO7 {
+		return // The pin is not configured as GPIO or GPIOHS.
+	}
+
+	// Configure pin.
+	kendryte.FPIOA.IO[uint8(p)].SetBits(kendryte.FPIOA_IO_OE_EN | kendryte.FPIOA_IO_IE_EN | kendryte.FPIOA_IO_ST | kendryte.FPIOA_IO_DS_Msk)
+
+	switch config.Mode {
+	case PinInput:
+		p.setFPIOAIOPull(fpioaPullNone)
+		input = true
+	case PinInputPullUp:
+		p.setFPIOAIOPull(fpioaPullUp)
+		input = true
+	case PinInputPullDown:
+		p.setFPIOAIOPull(fpioaPullDown)
+		input = true
+	case PinOutput:
+		p.setFPIOAIOPull(fpioaPullNone)
+		input = false
+	}
+
+	if f >= FUNC_GPIO0 && f <= FUNC_GPIO7 {
+		// Converts the IO pin number in the effective GPIO number (based on the FPIOA function).
+		gpioPin := uint8(f - FUNC_GPIO0)
+
+		if input {
+			kendryte.GPIO.DIRECTION.ClearBits(1 << gpioPin)
+		} else {
+			kendryte.GPIO.DIRECTION.SetBits(1 << gpioPin)
+		}
+	} else if f >= FUNC_GPIOHS0 && f <= FUNC_GPIOHS31 {
+		// Converts the IO pin number in the effective GPIOHS number (based on the FPIOA function).
+		gpioPin := uint8(f - FUNC_GPIOHS0)
+
+		if input {
+			kendryte.GPIOHS.INPUT_EN.SetBits(1 << gpioPin)
+			kendryte.GPIOHS.OUTPUT_EN.ClearBits(1 << gpioPin)
+		} else {
+			kendryte.GPIOHS.OUTPUT_EN.SetBits(1 << gpioPin)
+			kendryte.GPIOHS.INPUT_EN.ClearBits(1 << gpioPin)
+		}
+	}
+}
+
+// Set the pin to high or low.
+func (p Pin) Set(high bool) {
+
+	// Check if the current pin's FPIOA function is either GPIO or GPIOHS.
+	f := p.FPIOAFunction()
+	if f < FUNC_GPIOHS0 || f > FUNC_GPIO7 {
+		return // The pin is not configured as GPIO or GPIOHS.
```

Comment:
This check is reasonable however (although it's also the caller's responsibility to not call `Set` on a pin that is not configured as GPIO pin).